### PR TITLE
alter_table: Update `alter_table_desc` API to use Persist's `compare_and_evolve_schema(...)`

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR updates the existing `alter_table_desc(...)` API previously introduced in https://github.com/MaterializeInc/materialize/pull/28633 to use the new schema mechanisms introduced by Persist.

* We now require an expected `SchemaId` when altering the `RelationDesc` for a table, this will be provided by the Adapter.
* `StorageController::alter_table_desc` calls into Persist's `compare_and_evolve_schema(...)` which will register the new write schema.

Further this PR also moves Persist's `SchemaId` type out of `persist-client` and into `persist-types` since it will be used by crates at the Adapter layer. This felt like the right move to me but folks should feel free to push back!

Note: The exact protocol for how the Adapter will call `alter_table_desc(...)`, specifically how we handle crash recovery, is still a bit hand-wavy but IMO this API for `alter_table_desc(...)` is the right one. Happy to chat with folks about this if they're curious, I'm intending to make the change in a stacked PR.

### Motivation

Progress towards https://github.com/MaterializeInc/materialize/issues/28082

### Tips for reviewer

This PR is split into separate commits which can be reviewed independently:

1. Updates to `StorageController::alter_table_desc`
2. Move `SchemaId` to `persist-types`, this is pure code movement.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/storage-controller/src/lib.rs [hunk 1]
```diff
@@ -1061,29 +1063,57 @@ where
...
+            storage_collections.alter_table_desc(table_id, new_desc.clone())?;
             // StorageCollections was successfully updated, now we can update our
             // in-memory state.
             collection.collection_metadata.relation_desc = new_desc.clone();
 
             collection.collection_metadata.data_shard
         };
 
+        let diagnostics = Diagnostics {
+            shard_name: shard_id.to_string(),
```
Comment (danhhz) [inline src/storage-controller/src/lib.rs:1076]: This should be the global_id instead

**Review by danhhz — APPROVED** [review]
LGTM!

> Further this PR also moves Persist's SchemaId type out of persist-client and into persist-types since it will be used by crates at the Adapter layer. This felt like the right move to me but folks should feel free to push back!

So far, the distinction has been that persist-types is the things you'd need to impl `Codec` (i.e. mz_repr depends on -types but not -client). Plus we moved in ShardId at some point for some reason I don't recall offhand. SchemaId doesn't quite fall into that bucket, but I'm fine with moving it in anyway.

**Review by jkosh44 — APPROVED** [review]

File: src/storage-controller/src/lib.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/storage-controller/src/lib.rs]: Updated!

**Comment by ParkMyCar** [comment]
> So far, the distinction has been that persist-types is the things you'd need to impl `Codec` (i.e. mz_repr depends on -types but not -client). Plus we moved in ShardId at some point for some reason I don't recall offhand. SchemaId doesn't quite fall into that bucket, but I'm fine with moving it in anyway.

Ahh I see! I do what to stick with the existing distinction, but `mz_repr` is exactly where I started using `SchemaId` because it's where `RelationVersion` lives and I wanted to bridge the two

**Comment by danhhz** [comment]
Yep sounds totally reasonable

**Pull request merged**